### PR TITLE
Twitcher/Magpie upgrade and Cache settings off

### PR DESCRIPTION
**Pull request by fmigneault**
## Overview

Applies the same changes as #174 (cache settings) but disable Twitcher caching (Magpie caching already disabled) 

This is in order to validate that cache settings are applied and tests suite can run successfully with cache disabled, 
to isolate that sporadic errors seen in #174 are related to enabled cache.

## Changes

**Non-breaking changes**
- Added cache settings of `MagpieAdapter` through Twitcher (intended to improve response times, but disabled in this PR).
- Update Magpie/Twitcher version 3.14.0

**Breaking changes**
- Unique user email in Magpie is enforced, see changelog for how to find them and update the conflicted values before the upgrade.  Otherwise the upgrade will break.

## Related Issue / Discussion

- Undo resolution of [DAC-296 Twitcher performance issue](https://www.crim.ca/jira/browse/DAC-296) since caching now disabled
- Undo resolution of bird-house/twitcher#97 since cache becomes disabled 
- Based on top of #174 with cache-settings fixes
- Related new stress test https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/74

### CODE REVIEW

**Comment by fmigneault** [comment]
run tests

**Comment by fmigneault** [comment]
run tests

**Comment by fmigneault** [comment]
@tlvu  @dbyrns  @matprov 
Tests seem to behave correctly over a few runs, so I believe it is safe to update Magpie/Twitcher that enforce `cache.enabled=false`. 
I can let you guys integrate this when desired. 
I will look into caching issues to re-enable it and take advantage of speedups after my return from vacations (start of August).

**Review by dbyrns — APPROVED** [review]
LGTM

File: CHANGES.md
Comment (matprov) [inline CHANGES.md:19]: typo in twitcher

File: CHANGES.md
Comment (matprov) [inline CHANGES.md:25]: `sqlalchemy.engine`

File: CHANGES.md
Comment (matprov) [inline CHANGES.md:23]: rephrase "provide relevant logging details provided by `MagpieAdapter` it employs"

**Review by matprov — APPROVED** [review]

**Review by tlvu — APPROVED** [review]
LGTM as well.  Will upgrade on my test VM and report back.  No DB migration?

**Comment by tlvu** [comment]
Oh I see 
```
[2021-07-21 16:46:44,776] INFO       [MainThread][magpie.db] Running database migration (as required)...
[2021-07-21 16:46:44,776] INFO       [MainThread][magpie.db] Using file '/opt/local/src/magpie/config/magpie.ini' for migration.
[2021-07-21 16:46:44,850] WARNING    [MainThread][magpie.utils] Constant could not be found: MAGPIE_DB_URL (using default: None)
[2021-07-21 16:46:44,851] INFO       [MainThread][magpie.db] Using composed settings 'MAGPIE_POSTGRES_<>' for database connection.
[2021-07-21 16:46:44,902] INFO       [MainThread][alembic.runtime.migration] Context impl PostgresqlImpl.
[2021-07-21 16:46:44,902] INFO       [MainThread][alembic.runtime.migration] Will assume transactional DDL.
[2021-07-21 16:46:44,949] INFO       [MainThread][alembic.runtime.migration] Running upgrade 954a9d7fe740 -> dea413e13a8a, Pending Users and statuses
[2021-07-21 16:46:45,132] INFO       [MainThread][alembic.runtime.migration] Running upgrade dea413e13a8a -> 00c617174e54, Unique user emails.
[2021-07-21 16:46:45,143] INFO       [MainThread][alembic.runtime.migration] Running upgrade 00c617174e54 -> 35e98bdc8aed, Create a column to allow TokenOperation reference to UserPending entry.
```

So there are DB changes?  The `Unique user emails.` one might break something.  Will retry the upgrade using prod DB.

**Comment by tlvu** [comment]
Indeed, upgrade do not work on our Prod DB:
```
[2021-07-21 17:47:02,798] INFO       [MainThread][alembic.runtime.migration] Running upgrade dea413e13a8a -> 00c617174e54, Unique user emails.        
[2021-07-21 17:47:02,825] WARNING    [MainThread][magpie.db] Database migration failed [IntegrityError('(psycopg2.errors.UniqueViolation) could not c$
eate unique index "uq_users_email"\nDETAIL:  Key (email)=(logan.travis@ouranos.ca) is duplicated.\n')]. Retrying... (1/5)                     
Traceback (most recent call last):                                                                                                                    
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1277, in _execute_context                                             
    cursor, statement, parameters, context                                                                                                            
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute                                                 
    cursor.execute(statement, parameters)                                                                                                             
psycopg2.errors.UniqueViolation: could not create unique index "uq_users_email"                                                               
DETAIL:  Key (email)=(logan.travis@ouranos.ca) is duplicated.
```

What SQL query will list me all the duplicate emails (there might be more than 1 user).

This is a backward-incompatible change, please mention it in the changelog.

**Comment by dbyrns** [comment]
Sorry @tlvu to hear that. Please note that @fmigneault is on vacation until the end of next week.

**Review by tlvu — CHANGES_REQUESTED** [review]
This PR is not backward-compatible transparent upgrade.  Please document SQL query to find all duplicate emails in changelog.  All users upgrading to this PR will first have to inspect their DB to avoid upgrade failure.

**Comment by tlvu** [comment]
Found the query to find users with duplicate email here `echo "select email,user_name from users" | docker exec -i postgres-magpie psql -U $POSTGRES_MAGPIE_USERNAME magpiedb | sort > /tmp/magpie_users.txt`.

Will retry upgrade on my Dev VM with Prod DB with duplicate email fixed.

**Comment by tlvu** [comment]
Upgrade works on my dev VM after I manually change all the duplicated emails (we had 5 users with duplicated emails).

I guess I can merge this PR next week and let Francis update all the PR feedbacks in a separate PR once he is back from vacation.

**Comment by dbyrns** [comment]
Given that merging does not equal deploying in prod I'm not against it, but if you're to update the prod take into consideration that there is still 1 week before Francis com back from vacation. It's a compromise between having low latency requests (statu quo with caching) and fixing unreliable requests (with this PR).

**Comment by tlvu** [comment]
> Given that merging does not equal deploying in prod I'm not against it, but if you're to update the prod take into consideration that there is still 1 week before Francis com back from vacation. It's a compromise between having low latency requests (statu quo with caching) and fixing unreliable requests (with this PR).

I was going to merge to actually deploy this to prod to stop all those unreliable requests (many Jenkins nightlies are failing).  But you probably convinced me it's safer to wait for Francis.  1 week is not too long since we had those unreliable behavior for much longer than 1 week already.

So I'll leave this PR open for Francis to implement all the feedbacks.

**Comment by fmigneault** [comment]
The detail about the email-duplicate rule is explained here: 

https://github.com/Ouranosinc/Magpie/blob/master/CHANGES.rst#3130-2021-06-29
(see first item under "Bug Fixes") 
It is quite an important security issue.

**Comment by tlvu** [comment]
> The detail about the email-duplicate rule is explained here:
> 
> https://github.com/Ouranosinc/Magpie/blob/master/CHANGES.rst#3130-2021-06-29
> (see first item under "Bug Fixes")
> It is quite an important security issue.

Agreed the fix is useful.  But this is the kind of fix that will break a seamless upgrade.

For the future, I propose that Magpie changelog has a "backward incompatible fixes" section to make it stand out and document any manual steps required (ex in this case how to query the DB to list duplicate email and then use the Magpie UI to update those duplicate email) to prep the instance before upgrade.  And in the changelog here, list the change in "Breaking changes" section of this PR template and reference the corresponding manual steps in Magpie documentation changelog.

Anyways, very happy with this upgrade, we are looking to enable the `MAGPIE_USER_REGISTRATION_ENABLED` flag once we confirm the intermittent issue is fixed by cache off setting.

I guess cache can be turned back on in a subsequent release once enough testing?

**Comment by fmigneault** [comment]
@tlvu 
I added a note in e4c1c91 about the email change introduced by Magpie with the command you found.

**Review by tlvu — APPROVED** [review]
LGTM, merge when ready.

**Comment by tlvu** [comment]
As usual, copy paste the PR description into the merge commit description.

**Comment by tlvu** [comment]
I updated the PR description to note the breaking change with the unique email constrain.  If my wording is wrong, please correct.

**Comment by tlvu** [comment]
Not running bumpversion before merging?  In another PR?  Do I tag this PR 1.14.0 now (bumping minor since backward incompat change) and you do the bumpversion in a separate PR?

**Comment by fmigneault** [comment]
I'll do another PR with the bump. Forgot to do it.

**Comment by tlvu** [comment]
Ouranos product has been upgraded.  The upgrade went well.

**Comment by tlvu** [comment]
Jenkins run on Ouranos production only known intermittent error (ESGF login and `WMS_example.ipynb` broken due to https://neowms.sci.gsfc.nasa.gov/wms/wms) so I'd say the upgrade is a success, thanks everyone !

We'll keep an eye on the stress-test.ipynb for a few days then we can look at enabling `MAGPIE_USER_REGISTRATION_ENABLED`.

**Comment by fmigneault** [comment]
@tlvu 
When you get back from vacation, we can start looking into re-enabling caching and solving any side-effect if issues are found. 
We will need to make sure it works correctly both on CRIM and Ouranos instances.
see PR #188

**Comment by tlvu** [comment]
@fmigneault note I found some new intermittent error with cache off, see https://github.com/Ouranosinc/Magpie/issues/433#issuecomment-891893418 and Hirondelle should be running the last Magpie with cache off so you can try reproducing it on your side as well, see https://github.com/Ouranosinc/Magpie/issues/433#issuecomment-891897568

**Pull request merged**
